Added ability to poll for container stats instead of constant stream

jcbowman opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/aws/amazon-ecs-agent/blob/master/CONTRIBUTING.md

Please provide the following information:
-->

### Summary
Allows for container stat polling instead of forcing constant stream. This allows for a significant decrease in CPU usage when a lot of containers are deployed.

### Implementation details
There's 2 environment configuration variables that allow you to disable stream (default is to leave streaming turned on) and specify the polling interval (default is 15 seconds).

### Testing
This was tested by replacing the ecs-agent container on a ECS host with over 150 containers. We've been using the polling code for the past month in both our prod and nonprod environments.

- [x] Builds on Linux (`make release`)
- [x] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [x] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [ ] Integration tests on Linux (`make run-integ-tests`) pass
- [ ] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [ ] Functional tests on Linux (`make run-functional-tests`) pass
- [ ] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes: <!-- yes|no -->

### Description for the changelog
Feature - Allow for Container Stat Polling

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
 

fixes https://github.com/aws/amazon-ecs-agent/issues/1422

fenxiong commented:
Hi @jcbowman ,
Thanks for creating this PR! We will take a look at it. Also, can you change the base branch to dev, since that's where we will merge PR into?

Thanks,
Feng

jcbowman commented:
Thanks for the guidance. I changed the base branch to dev and squashed my changes. For some reason the builds started failing after pointing to dev. At a first glance the failing test doesn't appear related to these changes - I'll try to take a deeper look at it soon.

petderek commented:
@jcbowman It looks like it was just a transient failure -- so no worries!

boynux commented:
Is there any ETA for this change to be merged and rolled out?

richardpen commented on agent/config/config_unix.go:
```diff
@@ -49,6 +49,8 @@ func DefaultConfig() Config {
 		DisableMetrics:              false,
 		ReservedMemory:              0,
 		AvailableLoggingDrivers:     []dockerclient.LoggingDriver{dockerclient.JSONFileDriver, dockerclient.NoneDriver},
+		PollMetrics:                 false,
```
Can you modify the test in `config_test.go` to cover this?  Specifically, the following test:
https://github.com/aws/amazon-ecs-agent/blob/331f018333c70fe408f20028a38a3a28b03636dc/agent/config/config_test.go#L68

and add a test like this to test the default behavior:
https://github.com/aws/amazon-ecs-agent/blob/331f018333c70fe408f20028a38a3a28b03636dc/agent/config/config_test.go#L184

jcbowman replied:
Yes. Updated. I added tests to for testing the max and min poll interval also.

richardpen commented on agent/dockerclient/dockerapi/docker_client.go:
```diff
@@ -1244,24 +1245,53 @@ func (dg *dockerGoClient) Stats(id string, ctx context.Context) (<-chan *docker.
-			seelog.Infof("DockerGoClient: Unable to retrieve stats for container %s: %v",
-				id, statsErr)
+	if !dg.config.PollMetrics {
+		options := docker.StatsOptions{
```
Can you add a log line to indicate the behavior, same for the `else` case?

jcbowman replied:
Updated.

I added a log for 
"DockerGoClient: Starting to Poll for metrics"
or
"DockerGoClient: Starting to Stream for metrics"

depending on the situation.

richardpen commented on agent/dockerclient/dockerapi/docker_client.go:
```diff
@@ -1244,24 +1245,53 @@ func (dg *dockerGoClient) Stats(id string, ctx context.Context) (<-chan *docker.
+	} else {
+
+		//Sleep for a random period time up to the polling interval. This will help make containers ask for stats at different times
+		time.Sleep(time.Second * time.Duration(rand.Intn(int(dg.config.PollingMetricsWaitDuration.Seconds()))))
```
I don't think this is necessary, as the containers are not started at the same time.

jcbowman replied:
I don't think it'll happen a lot - the situation where I saw this come up is if the ECS service aws stopped and started when there were already a lot of containers on a server.

richardpen commented on agent/dockerclient/dockerapi/docker_client.go:
```diff
@@ -1244,24 +1245,53 @@ func (dg *dockerGoClient) Stats(id string, ctx context.Context) (<-chan *docker.
+					returnStats <- dockerStats
+				}
+
+				time.Sleep(dg.config.PollingMetricsWaitDuration)
```
Instead of using `sleep`, you can use a `timer` here.

jcbowman replied:
Converted to use timer

jcbowman commented on agent/config/config.go:
```diff
@@ -77,6 +81,14 @@ const (
+
+	// maximumPollingMetricsWaitDuration specifies the maximum duration to wait before polling for new stats
+	// from docker. This is only used when PollMetrics is set to true
+	maximumPollingMetricsWaitDuration = 60 * time.Second
```
Yeah, sure thing. Updated.

jcbowman commented on agent/config/config.go:
```diff
@@ -271,6 +283,11 @@ func (cfg *Config) validateAndOverrideBounds() error {
 	}
 
+	if cfg.PollingMetricsWaitDuration < minimumPollingMetricsWaitDuration || cfg.PollingMetricsWaitDuration > maximumPollingMetricsWaitDuration {
+		seelog.Warnf("Invalid value for polling metrics wait duration, will be overridden with the default value: %s. Parsed value: %v, minimum value: %v, maximum value: %v.", DefaultPollingMetricsWaitDuration.String(), cfg.PollingMetricsWaitDuration, minimumPollingMetricsWaitDuration, maximumPollingMetricsWaitDuration)
```
I updated these to only run if PollMetrics is true and split this into two different checks

jcbowman commented:
@richardpen I think I've addressed the PR comments. I squashed my commits to hopefully make merging easier. Let me know if there is anything else I should address.

adnxn commented:
> I've also deployed a version with the recent changes to one of our ECS servers - CPU is staying low and stats are showing up in CloudWatch.

Oh nice, this is great to hear.

@jcbowman, @boynux: We'd like to get this PR moving. Please address the branch conflicts and also unit tests appear to be failing. You can validate unit tests locally with `make test` target.

yunhee-l commented:
closing this in favor of https://github.com/aws/amazon-ecs-agent/pull/1646, please track the progress of this change through the new PR
